PowerPointLabs/PowerPointLabs/ActionFramework/Common/Extension/PPLClipboard.cs

```diff
@@ -0,0 +1,204 @@
+            });
+        }
+
+        public List<object> LoadClipboardObjects()
+        {
+            return LockIfNeeded(() =>
+            {
+                List<object> result = new List<object>();
+                if (Clipboard.ContainsImage())
+                {
+                    result.Add(Clipboard.GetImage());
+                }
+                if (Clipboard.ContainsFileDropList())
+                {
+                    result.Add(Clipboard.GetFileDropList());
+                }
+                if (Clipboard.ContainsText())
+                {
+                    result.Add(Clipboard.GetText());
+                }
+                return result;
+            });
+        }
+
+        public void LockAndRelease(Action action)
+        {
+            LockAndRelease<object>(() =>
+            {
+                action();
+                return null;
+            });
+        }
+
+        public TResult LockAndRelease<TResult>(Func<TResult> action)
+        {
+            LockClipboard();
+            try
+            {
+                return action();
+            }
+            catch (Exception e)
+            {
+                throw e;
+            }
+            finally
+            {
+                ReleaseClipboard();
+            }
+        }
+
+        public void LockClipboard()
+        {
+            if (IsLocked) { throw new Exception("Clipboard is not released before locking!"); }
+            // wait to lock the clipboard
+            while (!IsClipboardFree() && !OpenClipboard(_parentWindow))
+            {
+                if (!AutoDismiss)
+                {
+                    MessageBox.Show("Another application is currently using the clipboard. Please come back later and try again", "Retry", MessageBoxButton.OK);
+                }
+            }
+            IsLocked = true;
+        }
+
+        public void ReleaseClipboard()
+        {
+            if (!IsLocked) { return; }
+            CloseClipboard();
+            IsLocked = false;
+        }
+
+        public static void Init(IntPtr parentWindow, bool autoDismiss = false)
+        {
+            if (_instance == null)
+            {
+                _instance = new PPLClipboard(parentWindow, autoDismiss);
+            }
+        }
+
+        public void Teardown()
+        {
+            ReleaseClipboard();
+            _instance = null;
+        }
+
+        // not working stably
+        public void SaveClipboard()
+        {
+            lDataObject = Clipboard.GetDataObject();
+            string[] lFormats = lDataObject.GetFormats(false);
+            lBackup.Clear();
+            /*
+            foreach (string lFormat in lFormats)
+            {
+                if (lDataObject.GetDataPresent(lFormat))
+                {
+                    lBackup.Add(lFormat, lDataObject.GetData(lFormat, false));
+                }
+            }
+            */
+            //Clipboard.Clear(); // to prevent any weird data from showing up
```

Comment:
Any reason for the commented code?

Reply:
the code is suggested by an online post but seems to break the clipboard